Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertRebar.cs

```diff
@@ -1,175 +1,332 @@
-                _open.Add(iterator.Current as DB.Curve);
-            }
-            break;
-          default:
-            _open.Add(array.get_Item(0));
-            break;
-        }
+        return null;
       }
 
-      open = _open; closed = _closed;
-    }
+      // get type
+      var type = revitRebar.Document.GetElement(revitRebar.GetTypeId()) as ElementType;
 
-    private Base RebarToSpeckle(DB.Structure.Rebar revitRebar)
-    {
-      // get rebar centerline curves using transform
-      var bars = revitRebar.GetCenterlineCurves(true, true, true, MultiplanarOption.IncludeOnlyPlanarCurves, revitRebar.NumberOfBarPositions - 1);
-      var curves = new List<ICurve>();
+      // get the rebar shape
+      var revitShape = revitRebar.Document.GetElement(revitRebar.GetShapeId()) as DB.Structure.RebarShape;
+      RevitRebarShape speckleShape = RebarShapeToSpeckle(revitShape);
+#if REVIT2020 || REVIT2021
+      speckleShape.barDiameter = revitRebar.GetBendData().BarDiameter;
+#else
+      speckleShape.barDiameter = revitRebar.GetBendData().BarModelDiameter;
+#endif
 
+      // get the rebar hooks
+      DB.ElementId revitStartHookId = revitRebar.GetHookTypeId(0);
+      RevitRebarHook speckleStartHook = null;
+      double hookBendRadius = revitRebar.GetBendData().HookBendRadius;
+      if (revitStartHookId != ElementId.InvalidElementId)
+      {
+        var revitStartHook = revitRebar.Document.GetElement(revitStartHookId) as RebarHookType;
+        speckleStartHook = RebarHookToSpeckle(
+          revitStartHook,
+          revitRebar.GetHookOrientation(0).ToString(),
+          hookBendRadius
+        );
+      }
+      DB.ElementId revitEndHookId = revitRebar.GetHookTypeId(1);
+      RevitRebarHook speckleEndHook = null;
+      if (revitEndHookId != ElementId.InvalidElementId)
+      {
+        var revitEndHook = revitRebar.Document.GetElement(revitEndHookId) as RebarHookType;
+        speckleEndHook = RebarHookToSpeckle(revitEndHook, revitRebar.GetHookOrientation(1).ToString(), hookBendRadius);
+      }
+
+      // get the layout rule - this determines exceptions that may be thrown by accessing invalid props
+      bool isSingleLayout = revitRebar.LayoutRule == RebarLayoutRule.Single;
+
+      // get centerline curves for display value
       RebarShapeDrivenAccessor accessor = null;
       if (revitRebar.IsRebarShapeDriven())
         accessor = revitRebar.GetShapeDrivenAccessor();
+      // .GetCenterlineCurves() always returns the bar in the first position (even if it is excluded) for shape driven rebar groups
+      IList<DB.Curve> firstPositionCurves = revitRebar.GetCenterlineCurves(
+        true,
+        false,
+        false,
+        MultiplanarOption.IncludeAllMultiplanarCurves,
+        0
+      );
+
+      var centerlines = new List<ICurve>();
+      for (int i = 0; i < revitRebar.NumberOfBarPositions; i++)
+      {
+        // skip end bars that are excluded
+        if (!isSingleLayout)
+        {
+          if (
```

Comment:
I'd prefer the () braces start and don't behave like {} in this context as I find it makes it harder to read the if statement, i.e. with the ) at the same indent and immediately before the next {

Do we need both If statements? presumably we can collapse this into one?

Reply:
>"I'd prefer the () braces start and don't behave like {} in this context"

This is an autoformat decision, to be changed later

And yes, could collapse into one but I chose not to because accessing the `.IncludeFirstBar`and `.IncludeLastBar` props will throw if this is a single layout rebar, could get lost with the `&&` operator